R/sparsevctrs.R

```diff
@@ -1,3 +1,79 @@
 is_sparse_matrix <- function(x) {
   methods::is(x, "sparseMatrix")
 }
+
+toggle_sparsity <- function(object, data) {
+  toggle_sparse <- "no"
+
+  if (allow_sparse(object$fit$actions$model$spec)) {
+    if ("recipe" %in% names(object$pre$actions)) {
+      est_sparsity <- recipes::.recipes_estimate_sparsity(
+        object$pre$actions$recipe$recipe
+      )
+    } else {
+      est_sparsity <- sparsevctrs::sparsity(data, 1000)
+    }
+
+    pred_log_fold <- pred_log_fold(
+      est_sparsity,
+      object$fit$actions$model$spec$engine,
+      nrow(data)
+    )
+    if (pred_log_fold > 0) {
+      toggle_sparse <- "yes"
+    }
+  }
+
+  object$pre$actions$recipe$recipe <- recipes::.recipes_toggle_sparse_args(
+    object$pre$actions$recipe$recipe,
+    choice = toggle_sparse
+  )
+  object
+}
+
+allow_sparse <- function(x) {
+  if (inherits(x, "model_fit")) {
+    x <- x$spec
+  }
+  res <- parsnip::get_from_env(paste0(class(x)[1], "_encoding"))
+  all(res$allow_sparse_x[res$engine == x$engine])
+}
+
+pred_log_fold <- function(sparsity, model, n_rows) {
```

Comment:
This is generated from a mars model, using `sparsity`, `model` and `n_rows` from https://github.com/tidymodels/benchmark-sparsity-threshold

Reply:
I'm not married to this specific model. I'm fine with fine-tuning a little more. but it works fine as is

Reply:
I'm not sure what the name `pred_log_fold()` refers to. Could this function instead return `pred_log_fold() > 0` and instead be called something like `should_use_sparsity()` or `sparsity_will_accelerate()`?

Reply:
Let's add a link to that repo as a comment above this function!

Reply:
To an untrained eye, this function looks like it only supports the model types shown in the facets from that graph. Could we add a couple lines of explanatory comments discussing, at a high level, what these coefficients represent (as in--a model fit on __) and what the "usual" code path will be?